File: HW1/jdryuDer.m
% James Yu @jdryu 
% HW01-Differences

x = linspace(-5,10,500);
y = @(x) sin(sin(1./(x.*(2-x))));
[dyf,dxf] = Der(y,x,'fd');
[dyb,dxb] = Der(y,x,'bd');
[dyc,dxc] = Der(y,x,'cd');
[dy2,dx2] = Der(y,x,'o2');
[dy3,dx3] = Der(y,x,'o3');
[dy4,dx4] = Der(y,x,'o4');

[ddy,ddx] = Der(y,x,'d2');

[dY,dX] = Der(Y,X,'data');

[ddY,ddX] = Der(Y,X,'d2');

figure
plot(x,y(x))
title('sin^2(1/(x*(2-x)))')
legend('y(x)','Location','north')

figure
plot(dxf,dyf,dxb,dyb,dxc,dyc,dx2,dy2,dx3,dy3,dx4,dy4)
title('derivative of sin^2(1/(x*(2-x)))')
legend('fd','bd','cd','o2','o3','o4','Location','east')

figure
plot(ddx,ddy)
title('second derivative of sin^2(1/(x*(2-x)))')
legend('d2','Location','north')

figure
plot(X,Y)
title('derdata')
legend('Y','Location','north')

figure
plot(dX,dY)
title('derivative of derdata')
legend('d1','Location','north')

figure
plot(ddX,ddY)
title('second derivative of derdata')
legend('d2','Location','north')

File: HW1/jdryuDer.py
from HW1.myder import *
import numpy as np
from math import sin
from matplotlib import pyplot as plot

x = np.arange(-5, 10, 0.15)
f = lambda x: sin(sin(1/(x*(2-x))))
y = []
for i in x:
    y.append(f(i))

dxf = x[1:]
dyf = forwardiff(f, -5, 10, 100)
dxb = x[:len(x)-1]
dyb = backwardiff(f, -5, 10, 100)
dxc = x
dyc = centraldiff(f, -5, 10, 100, 1)
dx2 = x[1:]
dy2 = centraldiff(f, -5, 10, 100, 2)
dx3 = x[1:]
dy3 = centraldiff(f, -5, 10, 100, 3)
dx4 = x[1:len(x)-1]
dy4 = centraldiff(f, -5, 10, 100, 4)

ddx = x[1:]
ddy = seconddiff(f, -5, 10, 100)

plot.plot(x, y)
plot.title('$sin^2(1/(x*(2-x)))$')
plot.xlabel("x")
plot.ylabel("y")
plot.legend('y(x)')
plot.show()
plot.close()

plot.plot(dxf, dyf, label='fd')
plot.plot(dxb, dyb, label='bd')
plot.plot(dxc, dyc, label='cd')
plot.plot(dx2, dy2, label='o2')
plot.plot(dx3, dy3, label='o3')
plot.plot(dx4, dy4, label='o4')
plot.title('derivatives of $sin^2(1/(x*(2-x)))$')
plot.xlabel("x")
plot.ylabel("$dy/dx$")
plot.legend()
plot.show()
plot.close()

plot.plot(ddx, ddy)
plot.title('second derivative of $sin^2(1/(x*(2-x)))$')
plot.xlabel("x")
plot.ylabel("$d^2y/d^2x$")
plot.show()
plot.close()
